packages/x-date-pickers/src/DateCalendar/DateCalendar.tsx

```diff
@@ -255,7 +259,9 @@ export const DateCalendar = React.forwardRef(function DateCalendar<TDate>(
     }
   }, [value]); // eslint-disable-line
 
-  const ownerState = props;
+  // we only need one prop resolved correctly, passing empty object shouldn't be a problem
+  const disableDayMargin = resolveComponentProps(slotProps?.day, {} as any)?.disableMargin;
```

Comment:
I'm not fond of this solution, do you have any alternative suggestions? 🤔 
We could maybe leave this case unhandled because I guess that a maximum 1% of users might have seen a need for `disableMargin` on `DatePicker`.
As for DateRangePicker, it's handled in a different way and I reckon that this prop is mostly used only through Range Calendar. 🤔 
Removing this would produce the following result on a date picker with `slotProps.day.disableMargin = true`:
![Screenshot 2023-05-17 at 20 16 49](https://github.com/mui/mui-x/assets/4941090/9263be82-2880-480e-b814-f440f7601179)

Reply:
What's happening here? Is this calendar including a margin around the number by default, but this is a screenshot with the margin set to 0?? 
Is there a use case for removing the margin?

Reply:
This is more of a technical review question.
The `DateCalendar` has a `disableMargin` prop, which strips the margin from the days and currently, I have a bit hacky solution for handling this in code so that the width would adapt/shrink to account for that.
I imagine that basically, no one is using this prop with `DatePicker` though and this prop was mostly intended to be used on `DateRangeCalendar`, where stripping `margin` is a must to maintain the highlighted range without gaps. 🤔 

That's why I was suggesting maybe dropping the hacky solution and leaving this visual edge-case if anyone is actually using `disableMargin` with `DatePicker`. it's not that it would introduce any visual regression... 🤷 
This is the current state without and with the prop. 😉 
![Screenshot 2023-05-18 at 08 54 33](https://github.com/mui/mui-x/assets/4941090/fac93cbe-61db-4236-b802-be8d1e0ca80e)